engine/common/soundlib/snd_ogg_opus.c

```diff
@@ -0,0 +1,281 @@
+} opus_streaming_ctx_t;
+
+static int OpusCallback_Read( void *datasource, byte *ptr, int nbytes )
+{
+	return OggFilestream_Read( ptr, 1, nbytes, datasource );
+}
+
+static opus_int64 OpusCallback_Tell( void *datasource )
+{
+	return OggFilestream_Tell( datasource );
+}
+
+static int FS_ReadOggOpus( void *datasource, byte *ptr, int nbytes )
+{
+	opus_streaming_ctx_t *ctx = (opus_streaming_ctx_t*)datasource;
+	return g_fsapi.Read( ctx->file, ptr, nbytes );
+}
+
+static int FS_SeekOggOpus( void *datasource, int64_t offset, int whence )
+{
+	opus_streaming_ctx_t *ctx = (opus_streaming_ctx_t*)datasource;
+	return g_fsapi.Seek( ctx->file, offset, whence );
+}
+
+static opus_int64 FS_TellOggOpus( void *datasource )
+{
+	opus_streaming_ctx_t *ctx = (opus_streaming_ctx_t*)datasource;
+	return g_fsapi.Tell( ctx->file );
+}
+
+static const OpusFileCallbacks op_callbacks_membuf = {
+	OpusCallback_Read,
+	OggFilestream_Seek,
+	OpusCallback_Tell,
+	NULL
+};
+
+static const OpusFileCallbacks op_callbacks_fs = {
+	FS_ReadOggOpus,
+	FS_SeekOggOpus,
+	FS_TellOggOpus,
+	NULL
+};
+
+/*
+=================================================================
+
+	Ogg Opus decompression & streaming
+
+=================================================================
+*/
+static void Sound_ScanOpusComments( const OggOpusFile *of )
+{
+	const char *value;
+	const OpusTags *tags = op_tags( of, -1 );
+	if( tags )
+	{	
+		if(( value = opus_tags_query( tags, "LOOPSTART", 0 ))) {
+			sound.loopstart = Q_atoi( value );
+			SetBits( sound.flags, SOUND_LOOPED );
+		}
+		else if(( value = opus_tags_query( tags, "LOOP_START", 0 ))) {
+			sound.loopstart = Q_atoi( value );
+			SetBits( sound.flags, SOUND_LOOPED );
+		}
+	}
+}
+
+qboolean Sound_LoadOggOpus( const char *name, const byte *buffer, fs_offset_t filesize )
+{
+	long ret;
+	ogg_filestream_t file;
+	OggOpusFile *of;
+	const OpusHead *opusHead;
+	size_t written = 0;
+
+	if( !buffer )
+		return false;
+
+	OggFilestream_Init( &file, name, buffer, filesize );
+	of = op_open_callbacks( &file, &op_callbacks_membuf, NULL, 0, NULL );
+	if( !of ) {
+		Con_DPrintf( S_ERROR "%s: failed to load (%s): file reading error\n", __func__, file.name );
+		return false;
+	}
+
+	opusHead = op_head( of, -1 );
+	if( opusHead->channel_count < 1 || opusHead->channel_count > 2 ) {
+		Con_DPrintf( S_ERROR "%s: failed to load (%s): unsuppored channels count\n", __func__, file.name );
+		return false;
+	}
+
+	// according to OggOpus specification, sound always encoded at 48kHz sample rate
+	// but this isn't a problem, engine can do resampling 
+	sound.channels = opusHead->channel_count;
+	sound.rate = 48000;
+	sound.width = 2; // always 16-bit PCM
+	sound.type = WF_PCMDATA;
+	sound.samples = op_pcm_total( of, -1 );
+	sound.size = sound.samples * sound.width * sound.channels;
+	sound.wav = (byte *)Mem_Calloc( host.soundpool, sound.size );
+	
+	// skip undesired samples before playing sound
+	if( op_pcm_seek( of, opusHead->pre_skip ) < 0 ) {
```

Comment:
Let's print error code for easier troubleshooting for us and users.

Is there a function that converts error code into a string? If so, it could be used rather than just throwing a cryptic number.